src/main/java/org/glite/authz/pep/pip/provider/oidc/impl/OidcProfileTokenImpl.java

```diff
@@ -0,0 +1,152 @@
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package org.glite.authz.pep.pip.provider.oidc.impl;
+
+import static org.glite.authz.common.profile.OidcProfileConstants.ID_ATTRIBUTE_OIDC_ACCESS_TOKEN;
+import static org.glite.authz.common.profile.OidcProfileConstants.ID_ATTRIBUTE_OIDC_CLIENTID;
+import static org.glite.authz.common.profile.OidcProfileConstants.ID_ATTRIBUTE_OIDC_GROUP;
+import static org.glite.authz.common.profile.OidcProfileConstants.ID_ATTRIBUTE_OIDC_ISSUER;
+import static org.glite.authz.common.profile.OidcProfileConstants.ID_ATTRIBUTE_OIDC_ORGANISATION;
+import static org.glite.authz.common.profile.OidcProfileConstants.ID_ATTRIBUTE_OIDC_SCOPE;
+import static org.glite.authz.common.profile.OidcProfileConstants.ID_ATTRIBUTE_OIDC_SUBJECT;
+import static org.glite.authz.common.profile.OidcProfileConstants.ID_ATTRIBUTE_OIDC_USER_ID;
+import static org.glite.authz.common.profile.OidcProfileConstants.ID_ATTRIBUTE_OIDC_USER_NAME;
+
+import java.util.Arrays;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import org.glite.authz.common.model.Attribute;
+import org.glite.authz.common.model.Request;
+import org.glite.authz.common.model.Subject;
+import org.glite.authz.oidc.client.model.TokenInfo;
+import org.glite.authz.pep.pip.provider.oidc.OidcProfileToken;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class OidcProfileTokenImpl implements OidcProfileToken {
+
+  private static final Logger LOG = LoggerFactory.getLogger(OidcProfileTokenImpl.class);
+  private static final String SCOPE_SEPARATOR = " ";
+
+  private final Set<String> oidcAttributes;
+
+  public OidcProfileTokenImpl() {
+
+    oidcAttributes = new LinkedHashSet<>();
+    oidcAttributes.add(ID_ATTRIBUTE_OIDC_ISSUER);
+    oidcAttributes.add(ID_ATTRIBUTE_OIDC_SUBJECT);
+    oidcAttributes.add(ID_ATTRIBUTE_OIDC_ORGANISATION);
+    oidcAttributes.add(ID_ATTRIBUTE_OIDC_CLIENTID);
+    oidcAttributes.add(ID_ATTRIBUTE_OIDC_USER_ID);
+    oidcAttributes.add(ID_ATTRIBUTE_OIDC_USER_NAME);
+    oidcAttributes.add(ID_ATTRIBUTE_OIDC_GROUP);
+    oidcAttributes.add(ID_ATTRIBUTE_OIDC_SCOPE);
+  }
+
+  @Override
+  public Optional<String> extractTokenFromRequest(Request request) {
+
+    Optional<String> accessToken = Optional.empty();
+    for (Subject sub : request.getSubjects()) {
+      for (Attribute attr : sub.getAttributes()) {
+        if (ID_ATTRIBUTE_OIDC_ACCESS_TOKEN.equals(attr.getId())) {
+          Set<Object> values = attr.getValues();
+          accessToken = Optional.of(values.iterator().next().toString());
```

Comment:
If values is empty a NPE will be raised here. Add a check that controls that the attribute has at least 1 value or otherwise raises an exception (with a meaningful error message)